### ISSUE TITLE
Problem launching a simple cluster (when use_public_ips is False)

### ISSUE THREAD

**Issue by wooters** [awslabs/cfncluster#302]
I'm seeing very long waits followed, eventually, by a failure when trying to create a simple test cluster. My `~/.cfncluster/config` looks like this:

```
[cluster simple]
master_instance_type = m5.large
compute_instance_type = m5.large
initial_queue_size = 1
max_queue_size = 2
maintain_initial_size = true
master_root_volume_size = 50
compute_root_volume_size = 50
vpc_settings = simple
key_name = mykey
ec2_iam_role = my-cluster-role
base_os = ubuntu1604

[aws]
aws_region_name = us-east-1

[global]
update_check = true
sanity_check = true
cluster_template = simple

[vpc simple]
master_subnet_id = subnet-xxxxxxxx
vpc_id = vpc-xxxxxxxx
use_public_ips = false
vpc_security_group_id = sg-xxxxxxxx
```
I create the cluster from the command line using `cfncluster create -nr testcluster`.

`Master` gets created in about 5 minutes. No apparent issues.

`Compute` gets created in about 5 minutes too - i.e. it shows up in the AWS console with the "Instance State" of "Running". However, after about 35 minutes, the `cfncluster create -nr testcluster` command fails with:

```
Status: cfncluster-testcluster - CREATE_FAILED
Cluster creation failed.  Failed events:
  - AWS::CloudFormation::Stack cfncluster-testcluster The following resource(s) failed to create: [ComputeFleet].
  - AWS::AutoScaling::AutoScalingGroup ComputeFleet Received 0 SUCCESS signal(s) out of 1.  Unable to satisfy 100% MinSuccessfulInstancesPercent requirement
```

I tried to diagnose the problem by looking at the logs on the `Compute` host. First, I ssh'd into `Master`, and from there, into `Compute`. The log file `/var/log/cfn-init.log` shows a bunch of `TimeoutError`s:

```
2018-01-19 17:03:04,329 [DEBUG] CloudFormation client initialized with endpoint https://cloudformation.us-east-1.amazonaws.com
2018-01-19 17:03:04,329 [DEBUG] Describing resource ComputeServerLaunchConfig in stack cfncluster-testcluster
2018-01-19 17:04:04,329 [WARNING] Timeout of 60 seconds breached
2018-01-19 17:04:04,330 [ERROR] Client-side timeout
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
    return f(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 231, in _timeout
    raise TimeoutError("Execution did not succeed after %s seconds" % duration)
TimeoutError
2018-01-19 17:04:04,331 [DEBUG] Sleeping for 0.208080 seconds before retrying
2018-01-19 17:04:04,540 [DEBUG] Describing resource ComputeServerLaunchConfig in stack cfncluster-testcluster
2018-01-19 17:05:04,540 [WARNING] Timeout of 60 seconds breached
2018-01-19 17:05:04,540 [ERROR] Client-side timeout
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
    return f(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 231, in _timeout
    raise TimeoutError("Execution did not succeed after %s seconds" % duration)
TimeoutError
<SNIP>
```

The end of `/var/log/cloud-init-output.log` has:

```
Cloud-init v. 17.1 running 'modules:config' at Fri, 19 Jan 2018 17:00:38 +0000. Up 29.35 seconds.
+ proxy=NONE
+ custom_cookbook=NONE
+ '[' NONE '!=' NONE ']'
+ proxy_args=
+ touch /tmp/proxy.sh
+ '[' NONE '!=' NONE ']'
+ '[' us-east-1 == us-east-1 ']'
+ s3_prefix=s3
+ cookbook_url=https://s3.amazonaws.com/cfncluster-resources-us-east-1/cookbooks/cfncluster-cookbook-1.4.1.tgz
+ export PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin
+ PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin
+ export cfncluster_version=cfncluster-1.4.2
+ cfncluster_version=cfncluster-1.4.2
+ export cookbook_version=cfncluster-cookbook-1.4.1
+ cookbook_version=cfncluster-cookbook-1.4.1
+ export chef_version=12.19.36
+ chef_version=12.19.36
+ export ridley_version=5.1.0
+ ridley_version=5.1.0
+ export berkshelf_version=5.6.4
+ berkshelf_version=5.6.4
+ '[' -f /opt/cfncluster/.bootstrapped ']'
++ cat /opt/cfncluster/.bootstrapped
+ installed_version=cfncluster-1.4.2
+ '[' cfncluster-1.4.2 '!=' cfncluster-1.4.2 ']'
+ mkdir /tmp/cookbooks
+ cd /tmp/cookbooks
++ cat /etc/chef/cfncluster-cookbook.tgz.date
cat: /etc/chef/cfncluster-cookbook.tgz.date: No such file or directory
+ curl -v -L -o /etc/chef/cfncluster-cookbook.tgz -z '' https://s3.amazonaws.com/cfncluster-resources-us-east-1/cookbooks/cfncluster-cookbook-1.4.1.tgz
Warning: Illegal date format for -z, --timecond (and not a file name).
Warning: Disabling time condition. See curl_getdate(3) for valid date syntax.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 52.216.19.51...
<REMOVED A BUNCH OF CURL CRUFT>
* connect to 52.216.19.51 port 443 failed: Connection timed out
* Failed to connect to s3.amazonaws.com port 443: Connection timed out
* Closing connection 0
curl: (7) Failed to connect to s3.amazonaws.com port 443: Connection timed out
+ tar -xzf /etc/chef/cfncluster-cookbook.tgz

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
+ cd /tmp
+ cfn-init -s cfncluster-testcluster -v -c default -r ComputeServerLaunchConfig --region us-east-1
Unknown error retrieving ComputeServerLaunchConfig
+ error_exit 'Failed to run cfn-init. If --norollback was specified, check /var/log/cfn-init.log and /var/log/cloud-init-output.log.'
+ cfn-signal --exit-code=1 '--reason=Failed to run cfn-init. If --norollback was specified, check /var/log/cfn-init.log and /var/log/cloud-init-output.log.' --stack=cfncluster-testcluster --resource=ComputeFleet --region=us-east-1
Unknown error signaling ComputeFleet
+ exit 1
Cloud-init v. 17.1 running 'modules:final' at Fri, 19 Jan 2018 17:00:53 +0000. Up 44.13 seconds.
2018-01-19 17:13:33,307 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/part-002 [1]
2018-01-19 17:13:33,317 - cc_scripts_user.py[WARNING]: Failed to run module scripts-user (scripts in /var/lib/cloud/instance/scripts)
2018-01-19 17:13:33,317 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_scripts_user.py'>) failed
Cloud-init v. 17.1 finished at Fri, 19 Jan 2018 17:13:33 +0000. Datasource DataSourceEc2Local.  Up 804.27 seconds
```

I'm not sure what is happening. It seems to be a problem communicating with `https://s3.amazonaws.com/cfncluster-resources-us-east-1`. Or maybe there is an issue with my configuration?

Thanks for any help you can provide.

**Comment by wooters** [comment]
Some more info. 

It does appear to be a problem with the `curl` command on `Compute`. When I run it by hand it times out:
```
Compute $ curl -v -L -o /tmp/cfncluster-cookbook.tgz https://s3.amazonaws.com/cfncluster-resources-us-east-1/cookbooks/
cfncluster-cookbook-1.4.1.tgz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 52.216.224.3...
  0     0    0     0    0     0      0      0 --:--:--  0:02:06 --:--:--     0* connect to 52.216.224.3 port 443 failed: Connection timed out
* Failed to connect to s3.amazonaws.com port 443: Connection timed out
* Closing connection 0
curl: (7) Failed to connect to s3.amazonaws.com port 443: Connection timed out
```

When I run the same command on `Master` it seems ok:

```
Master $ ll /tmp/cfncluster-cookbook.tgz
-rw-rw-r-- 1 ubuntu ubuntu 220264 Jan 19 18:02 /tmp/cfncluster-cookbook.tgz
```

**Comment by wooters** [comment]
Problem solved (I think). 

So, I noticed that even though I had set `use_public_ips = false` in the `[vpc simple]` section of my config, the `Master` host _was assigned_ a public IP address. The `Compute` was not. That seemed to be one of the only differences between `Master` and `Compute` so I commented out `use_public_ips = false` and then created a new cluster. This time it worked and the cluster was created in about 7 minutes!

I'm trying to improve my understanding of `cfncluster`. The [documentation for use_public_ips](https://cfncluster.readthedocs.io/en/latest/configuration.html#use-public-ips) says:
```
use_public_ips

Define whether or not to assign public IP addresses to EC2 instances.

Set to false if operating in a private VPC.

Defaults to true.
```
From this I assumed that **no** host would get a public IP address. 

So, I have two questions: 

1) should the documentation say that `use_public_ips` _only_ applies to compute hosts? Or is it a bug that a public IP was assigned to `Master`?

2) Is it expected that a host without a public IP _can't_ talk to `https://s3.amazonaws.com`? (I thought that if my vpc security group allowed _all_ outbound connections, I'd be ok...)

Apologies in advance if these are basic questions. I tried to get the answers on my own, but couldn't seem to find anything.

**Comment by jocherry** [comment]
Investigating this I see the same issue you are seeing.
Here are the answers to your specific questions:
1. Yes, the master host should always get a public IP address.  Otherwise there would be no way to connect to your cluster.  If that is unclear we can add a note to the docs to clear that up.
2. https://aws.amazon.com/blogs/aws/new-vpc-endpoint-for-amazon-s3/ should clarify, but there are ways to connect to S3 via a VPC.

The point remains that the functionality of this key is broken, and setting use_public_ips = false will not work.  
If you do not need to use this key, I would recommend you do not.  
If so, can I understand your use-case a bit more, why can't the compute nodes have public IPs?

We will work on a fix for our next release.

**Comment by jocherry** [comment]
After speaking with teammates it appears that use of this key implies prior knowledge of accessing a private VPC without public endpoints.  As in you can preset up a NAT or other mechanism for these hosts to talk to one another, and then restrict public access to your compute nodes.
NATs are a common way: https://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-nat-gateway.html

I will update the documentation to be clearer about pre-requisites for use.

**Comment by wooters** [comment]
Thanks very much for your responses!

I think my confusion about the `use_public_ips` option was a result of 1) my lack of understanding of all things AWS, and 2) the specifics of our company's setup (which maybe isn't common?) 

At our company, our devops gurus have created a VPN. When I (a somewhat devops-challenged researcher) join the VPN from my laptop, I can connect to our AWS instances via ssh using the instance's _private_ IP address (or via the instance's `Name` tag that appears in the AWS web console, which they've somehow added to the dns). So, I never use the public IP addresses of our instances. It all works great.

The first time I created a cluster using `cfncluster`, I saw that `Master` had ports 80 and 22 open to the world and I was alarmed. I read through the options for cfncluster and (mistakenly) concluded that `use_public_ips = false` would prevent `Master` (and the compute instances) from being seen by the outside world. 

But, in the end, it turned out that I didn't actually need `use_public_ips = false`. The solution that worked for me was to just set `vpc_security_group_id` to the security group corresponding to our VPN. After that, everything worked and `Master` didn't have ports 80 and 22 exposed to the world.
